### PR TITLE
[WIP] Refactor SchemaReader and object model

### PR DESCRIPTION
**Pull request by cristianoc72**
This PR starts from #61 , please read the discussion in that PR for further details.

- [X] Refactor Model classes, according to PR #61 
- [X] Drop loadMapping() completely, use getters and setters
- [x] Externalize finalize() methods from the model into somewhere else => $whatever->finalize($schema) 
- [x] Rewrite SchemaReader to make use of the new APIs and use ~~something different like symfony config component to support multiple schema formats~~ Symfony Config Component to load and process schemas. Supported schema formats are: xml, yaml, json, php. #69
- [x] Fix agnostic test suite (non-trivial operation: basically it means refactor managers and bilders to interact with new loaders and object model)

The new code is already PHP 7.1 strict.

### CODE REVIEW

**Comment by cristianoc72** [comment]
@gossi question:

let's suppose to have a schema with external schemas. After loading and parsing, I have one `Schema` object with a collection of external `Schema` objects. Before building model classes, the external schemas are merged into their parent. Am I right?

If yes:
since I load schemas with Symfony Config loaders, I build the object models starting from an array, no more from an xml file. So, I could load and merge external schemas into their parent at "process" time, before building the object model (merging arrays is easier the objects).
I mean:
1. load schema
2. process schema to normalize and validate it
3. loop into external schemas
        3.1  load external schema
        3.2 process external schema to normalize and validate it
5. merge external schemas into their parent (`array_merge_recursive` or some other function)
6. now build object model from the resulted array

The point 3 could be recursive (I could have an external schema into another external schema).

What do you think about?

**Comment by gossi** [comment]
There are two ways to handle external schemas:

1) Merge them in and generate them as you mentioned
2) Merge them in only for reference and only generate the models which are in the parent schema.

**Comment by cristianoc72** [comment]
Choose the second (I see, you prepared the `Schema` class for that).

Another question: I can't find `domain` in any schema nor into the documentation: should it be a tag under `database`? I.e.
```xml
<database name="my-db">
    <domain name="mydomain" type="........."  />
..........................................
```

**Comment by cristianoc72** [comment]
@marcj  @gossi any chance someone could review this huge PR? :grin:

**Comment by gossi** [comment]
I try when I return from work later today. I started this beast it's cool to see it jump in 😃

File: .travis.yml [hunk 1]
```diff
@@ -1,7 +1,8 @@
 language: php
 
 php:
-  - 7.1
+  - 7.2
```
Comment (gossi) [inline .travis.yml:4]: 👍

File: src/Propel/Generator/Behavior/Sortable/SortableBehavior.php [hunk 2]
```diff
@@ -175,7 +177,7 @@ public function activeRecordTraitBuilderModification(ActiveRecordTraitBuilder $b
 
     public function useScope()
```
Comment (gossi) [inline src/Propel/Generator/Behavior/Sortable/SortableBehavior.php]: return type... maybe similar methods in other behaviors that I missed, too.

File: src/Propel/Generator/Builder/DataModelBuilder.php [hunk 3]
```diff
@@ -505,7 +507,7 @@ public function getGeneratorConfig()
      *
      * @return string
      */
-    public function getBuildProperty($name)
+    public function getBuildProperty(string $name)
```
Comment (gossi) [inline src/Propel/Generator/Builder/DataModelBuilder.php:510]: return type?

File: src/Propel/Generator/Model/Domain.php [hunk 4]
```diff
@@ -203,29 +159,27 @@ public function getSize()
...
+        $this->size = $size;
     }
 
     /**
      * Replaces the size if the new value is not null.
      *
      * @param integer $size
      */
-    public function replaceSize($size)
+    public function replaceSize(int $size)
```
Comment (gossi) [inline src/Propel/Generator/Model/Domain.php]: similar to `setSize()`? Mark deprecated or remove?

File: src/Propel/Generator/Model/Domain.php [hunk 5]
```diff
@@ -245,31 +199,29 @@ public function setType($mappingType)
      *
      * @param string $mappingType
      */
-    public function replaceType($mappingType)
+    public function replaceType(string $mappingType)
```
Comment (gossi) [inline src/Propel/Generator/Model/Domain.php]: similar to `setType()`? Mark deprecated or remove?

File: src/Propel/Generator/Model/Entity.php [hunk 6]
```diff
@@ -1831,162 +1270,207 @@ public function hasAutoIncrement()
...
-    public function getDefaultAccessorVisibility()
+    public function isCrossRef(): bool
     {
-        return $this->defaultAccessorVisibility;
+        return $this->isCrossRef;
     }
 
     /**
-     * Sets the default mutator visibility.
+     * Alias for Entity::getIsCrossRef.
```
Comment (gossi) [inline src/Propel/Generator/Model/Entity.php]: This comment is wrong and also the method can possibly go away, in favor of `isCrossRef()` ?

File: src/Propel/Generator/Model/Field.php [hunk 7]
```diff
@@ -126,279 +146,77 @@ public function __construct($name = null, $type = null, $size = null)
...
-        return $this->parentEntity->getName() . '.' . strtoupper($this->getName());
-    }
-
-    /**
-     * Returns the column name.
-     *
-     * @return string
-     */
-    public function getName()
+    public function getFullyQualifiedName(): string
```
Comment (gossi) [inline src/Propel/Generator/Model/Field.php]: wasn't it `getFullName()` to stick with the naming scheme with the other methods?

File: src/Propel/Generator/Model/Field.php [hunk 8]
```diff
@@ -418,177 +236,21 @@ public function getSingularName()
...
-        $this->mutatorVisibility = $visibility;
-    }
-
     /**
      * Returns the full column constant name (e.g. EntityMapName::FIELD_COLUMN_NAME).
      *
      * @return string A column constant name for insertion into PHP code
      */
-    public function getFQConstantName()
+    public function getFQConstantName(): string
```
Comment (gossi) [inline src/Propel/Generator/Model/Field.php]: what about `getFullConstantName()`?

**Review by gossi — COMMENTED** [review]
I got until `src/Generator/Model/*` ... but it's very heavily loaded ;)

File: src/Propel/Generator/Builder/DataModelBuilder.php (hunk 3 above)
Comment (cristianoc72) [reply src/Propel/Generator/Builder/DataModelBuilder.php:510]: Mixed. I forgot to change the comment.

File: src/Propel/Generator/Behavior/Sortable/SortableBehavior.php (hunk 2 above)
Comment (cristianoc72) [reply src/Propel/Generator/Behavior/Sortable/SortableBehavior.php]: Fixed. useScope method in NestedSetBehavior has a correct return type.

File: src/Propel/Generator/Model/Domain.php (hunk 4 above)
Comment (cristianoc72) [reply src/Propel/Generator/Model/Domain.php]: Removed.

File: src/Propel/Generator/Model/Domain.php (hunk 5 above)
Comment (cristianoc72) [reply src/Propel/Generator/Model/Domain.php]: Removed.

File: src/Propel/Generator/Model/Entity.php (hunk 6 above)
Comment (cristianoc72) [reply src/Propel/Generator/Model/Entity.php]: Removed in favour of  `isCrossRef`

File: src/Propel/Generator/Model/Field.php (hunk 7 above)
Comment (cristianoc72) [reply src/Propel/Generator/Model/Field.php]: Fixed.

File: src/Propel/Generator/Model/Field.php (hunk 8 above)
Comment (cristianoc72) [reply src/Propel/Generator/Model/Field.php]: Done

**Comment by cristianoc72** [comment]
Well, this project is quiet dead. If no objection, I'ld like to merge this PR and try to move it over.

**Pull request merged**